clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1750,8 +1753,14 @@ private void subscribeInternal(Pattern pattern, Optional<ConsumerRebalanceListen
             throwIfNoAssignorsConfigured();
             log.info("Subscribed to pattern: '{}'", pattern);
             subscriptions.subscribe(pattern, listener);
-            updatePatternSubscription(metadata.fetch());
             metadata.requestUpdateForNewTopics();
+            Cluster cache = metadata.fetch();
+
+            while (cache == metadata.fetch()) {
+                log.info("Waiting for new metadata update");
+            }
```

Comment:
I guess this is with the purpose of ensuring we get metadata before we proceed to actually join the group with the resolved list of topic names. But the way it is I think is not conceptually right. With this we're only ensuring that there is a metadata, but that does not ensure that we have discovered any topic that matches the regex (I expect that in that case we would come out of the while, and the `SubscriptionChangeEvent` would lead to a first HB to join the group but with empty topic list, which would be an invalid first HB request).

I wonder if we could simply reuse the logic that we already have in `updatePatternSubscription`, that knows when topics matching the regex are discovered, and it's only at that point that we generate the event and join the group? We would end up with a much simpler logic here (only needing to `subscribe`, `updatePatternSubscription` and `requestUpdateForNewTopics` like it was before), and we would enqueue the subscription change event only when we actually discover topics ([here](https://github.com/apache/kafka/blob/f8ce7feebcede6c6da93c649413a64695bc8d4c1/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L1466)) . What do you think?

Reply:
@Phuc-Hong-Tran—why was this `while` loop added? Did you see issues in your testing, or were you preemptively wanting to make the code more robust?

Reply:
I'm wondering why we call `metadata.requestUpdateForNewTopics` here when that's already in `updatePatternSubscription`? 🤔

Reply:
@kirktrue, please see my comments below

Reply:
@lianetm Please correct me if I'm wrong, in't by default the `metadata.fetch()` doesn't return a null? With this assumption I decided to put the while loop there so that we can check if the consumer has received a newly updated metadata or not, other while we would still ended up with a invalid heartbeat request.

Reply:
@lianetm, I agree with the flow you suggested, however I'm not sure how can we confirm the fact that new topic(s) has been discovered after the `requestUpdateForNewTopics`

Reply:
hey @Phuc-Hong-Tran, when we call the `updatePatternSubscription`, we are passing the latest metadata already, so if we make it into the if [here](https://github.com/apache/kafka/blob/cf2874a2db845daaed80808ce42b50fc05584fdb/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L1466) it means that, with the latest metadata, we discovered new topics that made the subscription change. So only at that point we need to create the `SubscriptionChangeEvent` I would say (and yes, then we also call `requestUpdateForNewTopics`)

So echoing my first comment on this thread, seems to me that we shouldn't rely on any metadata object/version check here as it could not be accurate (removing [this](https://github.com/apache/kafka/blob/cf2874a2db845daaed80808ce42b50fc05584fdb/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L1757-L1761)), and we should just make sure that we send the `SubscriptionChangeEvent` only when we know that the subscription changed, which is inside the `updatePatternSubscription` [if](https://github.com/apache/kafka/blob/cf2874a2db845daaed80808ce42b50fc05584fdb/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java#L1465).

Makes sense?

Reply:
btw, the moment we move the event generation to the point where the subscription is updated inside the `updatePatternSubscription`, I expect you won't get the error you mentioned [here](https://github.com/apache/kafka/pull/15585#issuecomment-2021526537) that made you introduce the while loop:

> there is a race condition bug where the metadata is not updated but the heartbeat request is already created, but it lacks required info

Reply:
@lianetm, I understand, will fix it accordingly

Reply:
@lianetm, just to clarify though, do we expect to send a `SubscriptionChangeEvent` right after we call `subscribe(Pattern)` or we expect to send a `SubscriptionChangeEvent` in the next call of `poll`?